packages/plugin-ext-headless/src/hosted/node/plugin-host-headless.ts

```diff
@@ -0,0 +1,137 @@
+});
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const unhandledPromises: Promise<any>[] = [];
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+process.on('unhandledRejection', (reason: any, promise: Promise<any>) => {
+    unhandledPromises.push(promise);
+    setTimeout(() => {
+        const index = unhandledPromises.indexOf(promise);
+        if (index >= 0) {
+            promise.catch(err => {
+                unhandledPromises.splice(index, 1);
+                if (terminating && (ConnectionClosedError.is(err) || ConnectionClosedError.is(reason))) {
+                    // during termination it is expected that pending rpc request are rejected
+                    return;
+                }
+                console.error(banner, `Promise rejection not handled in one second: ${err} , reason: ${reason}`);
+                if (err && err.stack) {
+                    console.error(banner, `With stack trace: ${err.stack}`);
+                }
+            });
+        }
+    }, 1000);
+});
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+process.on('rejectionHandled', (promise: Promise<any>) => {
+    const index = unhandledPromises.indexOf(promise);
+    if (index >= 0) {
+        unhandledPromises.splice(index, 1);
+    }
+});
+
+let terminating = false;
+const channel = new IPCChannel();
+const rpc = new RPCProtocolImpl(channel);
+
+process.on('message', async (message: string) => {
+    if (terminating) {
+        return;
+    }
+    try {
+        const msg = JSON.parse(message);
+        if (ProcessTerminateMessage.is(msg)) {
+            terminating = true;
+            if (msg.stopTimeout) {
+                await Promise.race([
+                    pluginHostRPC.terminate(),
+                    new Promise(resolve => setTimeout(resolve, msg.stopTimeout))
+                ]);
+            } else {
+                await pluginHostRPC.terminate();
+            }
+            rpc.dispose();
+            if (process.send) {
+                process.send(JSON.stringify({ type: ProcessTerminatedMessage.TYPE }));
+            }
+
+        }
+    } catch (e) {
+        console.error(banner, e);
+    }
+});
+
+const container = new Container();
+container.bind(PluginContainerModuleLoader).toConstantValue((module: ContainerModule) => {
+    container.load(module);
+    const internalModule = module as InternalPluginContainerModule;
+    const pluginApiCache = internalModule.initializeApi?.(container);
+    if (pluginApiCache) {
+        // TODO
+        // pluginManager.onDidChange(() => {
+        //     if (!pluginManager.isDisposed) {
+        //         pluginApiCache.dispose();
+        //     }
+        // });
+    }
+});
+container.bind(RPCProtocol).toConstantValue(rpc);
+
+container.bind(AbstractPluginHostRPC).toService(HeadlessPluginHostRPC);
+container.bind(HeadlessPluginHostRPC).toSelf().inSingletonScope();
+container.bind(EnvExtImpl).to(EnvNodeExtImpl).inSingletonScope();
```

Comment:
Is there any particular reason we bind the environment extension and the localization extension against an actual implementation but only use dummies for the others, see  `HeadlessPluginManagerExtImpl`? Do we need those two for some special reason as opposed to the others? Could they also be

Reply:
Seems you didn't finish your thought, but I can answer in part, which is that my understanding is that the LocalizationExt is used for localization of strings in the package manifest, which I thought we would need. The `EnvExtImpl` I think maybe we don't need any longer as it was only used to provide some minimal capabilities of the default API object (such as getting environment variables) but we've since removed that default API. So I shall try excising the EnvExtImpl. Not sure about the localization.

Reply:
Ah, you are right! Thanks for the explanation. Localization of the package manifest may absolutely make sense! Would be great to document that.

And I believe how that should have finished is (but now I'm not sure that makes sense anymore):
> Do we need those two for some special reason as opposed to the others? Could they also be
... injected using the same mechanism the user would be using (PluginContainerModule? something else?).

Reply:
> ... injected using the same mechanism the user would be using (PluginContainerModule? something else?).

Yes, I definitely want to think about that. Makes perfect sense!